File: backend/src/resolvers/analytics.ts
import { Router } from "express";

import type { AnalyticsService } from "../services/analytics.js";

function groupBy(value: unknown) {
  return typeof value === "string" ? value : "country";
}

export function analyticsRoutes(deps: { analytics: AnalyticsService; today: () => string }) {
  const router = Router();

  router.get("/headcount", async (req, res, next) => {
    try {
      res.json(await deps.analytics.headcount(groupBy(req.query.groupBy)));
    } catch (error) {
      next(error);
    }
  });

  router.get("/avg-salary", async (req, res, next) => {
    try {
      const rows = await deps.analytics.avgSalary(groupBy(req.query.groupBy), deps.today());
      res.json(
        rows.map((row) => ({
          group: row.group,
          avg_salary_usd: row.avgSalaryUsd?.toFixed(2) ?? null,
          paid_headcount: row.paidHeadcount,
          headcount: row.headcount,
        })),
      );
    } catch (error) {
      next(error);
    }
  });

  router.get("/spend", async (req, res, next) => {
    try {
      const rows = await deps.analytics.spend(groupBy(req.query.groupBy), deps.today());
      res.json(
        rows.map((row) => ({
          group: row.group,
          spend_usd: row.spendUsd.toFixed(2),
          paid_headcount: row.paidHeadcount,
        })),
      );
    } catch (error) {
      next(error);
    }
  });

  return router;
}

File: backend-py/app/routers/analytics.py
from __future__ import annotations

from fastapi import APIRouter, Request

from app.jsonutil import money

router = APIRouter(prefix="/api/analytics")


def _group_by(value: str | None) -> str:
    return value if isinstance(value, str) else "country"


@router.get("/headcount")
def headcount(request: Request, groupBy: str | None = None):
    analytics = request.app.state.analytics
    return analytics.headcount(_group_by(groupBy))


@router.get("/avg-salary")
def avg_salary(request: Request, groupBy: str | None = None):
    analytics = request.app.state.analytics
    today = request.app.state.today
    rows = analytics.avg_salary(_group_by(groupBy), today())
    return [
        {
            "group": row["group"],
            "avg_salary_usd": money(row["avg_salary_usd"]) if row["avg_salary_usd"] is not None else None,
            "paid_headcount": row["paid_headcount"],
            "headcount": row["headcount"],
        }
        for row in rows
    ]


@router.get("/spend")
def spend(request: Request, groupBy: str | None = None):
    analytics = request.app.state.analytics
    today = request.app.state.today
    rows = analytics.spend(_group_by(groupBy), today())
    return [
        {
            "group": row["group"],
            "spend_usd": money(row["spend_usd"]),
            "paid_headcount": row["paid_headcount"],
        }
        for row in rows
    ]
